Jakartaee auth0 (#1332)
oktadeveloper/okta-blog: _source/_posts/2023-01-24-jakartaee-auth0.md

---
layout: blog_post

> title: "Open ID authentication with Jakarta EE 10 and Security 3.0"

Reviewer: ```suggestion
title: "OpenID Connect Authentication with Jakarta EE 10 and Security 3.0"
```

author: andrew-hughes
by: contractor
communities: [java,security]
description: "Use Jakarta EE 10 to build a secure Java web application using Open ID connect and Auth0."
tags: [java, jakartaee]
tweets:
- ""
- ""
- ""
image:
type: conversion
---

> ## Open ID authentication with Jakarta EE 10 and Security 3.0

Reviewer: ```suggestion
```
Not needed since the title is already above.

> Jakarta EE 10 includes a new authentication mechanism: Open ID Connect! This can be added to a Jakarta EE servlet using the new `@OpenIdAuthenticationMechanismDefinition` annotation. 

Reviewer: ```suggestion
Jakarta EE 10 includes a new authentication mechanism: OpenID Connect! This can be added to a Jakarta EE servlet using the new `@OpenIdAuthenticationMechanismDefinition` annotation. 
```

Reviewer: @moksamedia Please change `Open ID` to `OpenID` throughout. https://auth0.com/docs/authenticate/protocols/openid-connect-protocol

In this tutorial, you are going to see how to implement a web application with Open ID Connect (OIDC) authentication and use Auth0 as the OIDC provider. You are also going to see one way to secure an API and authenticate using JSON Web Tokens (JWTs). This will all be accomplished using Wildfly as the Jakarta EE runtime.

> This stack includes a lot of technologies. I'm going to introduce them briefly below. If you're comfortable with all those terms and just want to get to the code, **feel free to skip ahead to the requirements section**. 

Reviewer: todo: link to anchor below.

**Jakarta vs Java, EE vs SE**

> Jakarta EE is Jakarta Enterprise Edition. This was formerly Java EE. The name and framework packages were migrated when Oracle gave Java EE to the Eclipse Foundation because Oracle still has the rights to the Java brand and did not open-source absolutely everything that was in the `javax.*` namespace. Thus, Jakarta EE is the Eclipse-owned and now totally open-source Java EE (You may have recently had to change some packages from `javax` to `jakarta`. This is why.)

Reviewer: ```suggestion
Jakarta EE is Jakarta Enterprise Edition, formerly known as Java EE. The name and framework packages were migrated when Oracle gave Java EE to the Eclipse Foundation because Oracle still has the rights to the Java brand and did not open-source absolutely everything that was in the `javax.*` namespace. Thus, Jakarta EE is the Eclipse-owned and now totally open-source Java EE (You may have recently had to change some packages from `javax` to `jakarta`. This is why.)
```

You can check out [the docs for Jakarta EE 10 Security 3.0 OIDC here](https://jakarta.ee/specifications/security/3.0/jakarta-security-spec-3.0.html#openid-connect-annotation).

**Requirements**

Before you start, please make sure you have the following prerequisites installed (or install them now).

- [Java 17](https://adoptium.net/): or use [SDKMAN!](https://sdkman.io/) to manage and install multiple versions (the Jakarta EE spec says 11 and up is supported, but I wrote this tutorial assuming version 17)
- [Auth0 CLI](https://github.com/auth0/auth0-cli#installation): the Auth0 command-line interface
- [HTTPie](https://httpie.org/doc#installation): a simple tool for making HTTP requests from a Bash shell

**You will need a free Auth0 developer account** if you don't already have one. Go ahead and sign up for an Auth0 account using [their sign-up page](https://auth0.com/signup).

> Clone the tutorial from [the GitHub repository](need.a.link).

Reviewer: todo: create a repository and link to it

> ## Take a look at the build configuration and project dependencies

Reviewer: If you look at the preview on Netlify, you'll see this is quite long. You can add a TOC at the beginning to better visualize all the headlines. 

```
{% include toc.md %}
```

Author: OK. I removed some of it. Sometimes I get into the weeds a little.

I won't reproduce the entire `pom.xml` file here, but I want to point out a few things.

First, take a look at the dependencies. The only dependency required for Jakarta EE is the first one (`jakarta.jakartaee-api`). 

```xml
<dependencies>
    <dependency>
        <groupId>jakarta.platform</groupId>
        <artifactId>jakarta.jakartaee-api</artifactId>
        <version>${jakartaee-api.version}</version>
        <scope>provided</scope>
    </dependency>
</dependencies>
```

The WildFly plugin is included and configured in the block below.

```xml
<plugin>
    <groupId>org.wildfly.plugins</groupId>
    <artifactId>wildfly-maven-plugin</artifactId>
    <version>${wildfly-maven-plugin.version}</version>
    <configuration>
        <commands>
            <command>/subsystem=undertow/application-security-domain=other:write-attribute(name=integrated-jaspi, value=false)</command>
            <command>reload</command>
        </commands>
    </configuration>
</plugin>
```

> The docs for [the WildFly maven plugin are here](https://docs.wildfly.org/wildfly-maven-plugin/). Except for the cryptic `<command></command>` block, the plugin is pretty simple and easy to use. 

Reviewer: ```suggestion
The docs for [the WildFly Maven plugin are here](https://docs.wildfly.org/wildfly-maven-plugin/). Except for the cryptic `<command></command>` block, the plugin is pretty simple and easy to use. 
```

> It took a little digging to figure out, but the obscure command block *is* required, at least according to the experts I asked. It disables integrated JASPI in the server and instead delegates validation of credentials to a non-integrated ServerAuthModule. This allows for identities to be dynamically created instead of statically stored in an integrated security domain. Take a look at the [Elytron and Java EE Security section of the docs](https://docs.wildfly.org/26/WildFly_Elytron_Security.html#Elytron_and_Java_EE_Security) for more on this.

Reviewer: ```suggestion
It took a little digging to figure out, but the obscure command block *is* required, at least according to the experts I asked. It disables integrated JASPI in the server and instead delegates validation of credentials to a non-integrated `ServerAuthModule`. This allows for identities to be dynamically created instead of statically stored in an integrated security domain. Take a look at the [Elytron and Java EE Security section of the docs](https://docs.wildfly.org/26/WildFly_Elytron_Security.html#Elytron_and_Java_EE_Security) for more on this.
```

## Create an Auth0 API and OIDC application

If you have not already, install the [Auth0 CLI](https://github.com/auth0/auth0-cli) and run `auth0 login` in a terminal.

```bash
Waiting for the login to complete in the browser... done

 ▸    Successfully logged in.
 ▸    Tenant: dev-0xb84jzp.us.auth0.com
```

Take note of the domain listed as the tenet. This is your Auth0 domain. If you need to find it again later, you can use `auth0 tenants list`.

You need to create an API on Auth0. The Auth0 API is what exposes identity functionality for all authentication and authorization protocols, such as OpenID Connect and OAuth. **Without the API, Auth0 servers will return an opaque token that the Java application will not be able to verify.** This can lead to some cryptic error messages and some lost time. 

> Use the following command to create a custom Auth0 API named `myapi` with the identifier `http://my-api`.

Reviewer: This worked well in your last post, but I can't help but think using the default API might be better. The default is `https://{your-auth0-domain}/api/v2/`

Author: OK. I removed the custom API stuff.

```bash
auth0 apis create -n myapi --identifier http://my-api
```

Just press enter three times to accept the default values for scopes, token lifetime, and to allow offline access. The scopes here refer to custom scopes, not the standard scopes (email, profile, and openid) that you will need for OIDC and OAuth.

Now use the Auth0 CLI to create an OpenID Connect (OIDC) application.  From the project base directory, run the following.

```bash
auth0 apps create
```

Use the following values:

- **Name**: `javartaee-demo`

- **Description**: whatever you like, or leave blank
- **Type**: `Regular Web Application`
- **Callback URLs**: `http://localhost:8080/callback`
- **Allowed Logout URLs**: `http://localhost:8080`

The console output shows you the Auth0 domain and the OIDC client ID. However, you also need the client secret, which you have to get by logging into Auth0. Type the following:

```bash

> auth0 apps open

Reviewer: With the 1.0 version of the Auth0 CLI, you can use `--reveal-secrets` to see the client secret!

https://github.com/auth0/auth0-cli/issues/488#issuecomment-1404051864

```

Select the OIDC app (or client) you just created from the list. This will open the OIDC application on the Auth0 dashboard.

> {% img blog/jakartaee-auth0/oidc-application-auth0.png alt:"Auth0 OIDC App" width:"800" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/jakartaee-auth0/oidc-application-auth0.png alt:"Auth0 OIDC App" width:"600" %}{: .center-image }
```

Fill in the three values in `src/main/resources/openid.properties`. Replace the bracketed values with the values from the OIDC application page on the Auth0 dashboard.

```properties
issuerUri=<your-auth0-domain>
clientId=<your-client-id>

> clientSecret=<your-client-secret>ca

Reviewer: ```suggestion
clientSecret=<your-client-secret>
```

```

## Configure Roles on Auth0

> Open your [Auth0 developer dashboard](https://manage.auth0.com). You need to create a role, assign your user to that role, and create an action that will inject the roles into a custom claim in the JWT.

Reviewer: Is this possible from the CLI? I'm not sure as I haven't looked into it. It certainly would be cool if it was! Can you please check and see if 1.0 has it?

Under **User Management** click on **Roles**. Click the **Create Role** button. 

> {% img blog/jakartaee-auth0/auth0-create-role.png alt:"Auth0 Create Role" width:"1000" %}{: .center-image }

Reviewer: If you want full-width, I usually just remove the width attribute.

Select **Actions** from the left menu in the developer dashboard. Click on **Flows**. Select **Login**.

Add a new action by clicking on the **+** symbol to the right of **Add Action**. Select **Build Custom**.

Give the action a **Name**, such as `Add Roles`. Leave the other two values the same. Click **Create**.

{% img blog/jakartaee-auth0/auth0-create-action.png alt:"Auth0 Create Action" width:"600" %}{: .center-image }

Change the code for the action to the following.

```js
exports.onExecutePostLogin = async (event, api) => {
  const namespace = 'http://www.jakartaee.demo';
  if (event.authorization) {
    api.idToken.setCustomClaim('preferred_username', event.user.email);
    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
  }
}
```

Click on **Deploy**. 

Click on the **Add to flow** link in the popup window that slides in (if you miss this, you can find the new action under the custom action tab back in the flow panel).

Drag the **Add Roles** action over under the **Rules (legacy)** action. 

> {% img blog/jakartaee-auth0/auth0-action-flow.png alt:"Auth0 Action Flow" width:"600" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/jakartaee-auth0/auth0-action-flow.png alt:"Auth0 Action Flow" width:"400" %}{: .center-image }
```

Click **Apply** (top right of the panel).

> ## Take a look at the ProtectedServlet and OIDC flow

Reviewer: This can probably be a better headline. Adding the TOC should help visualize.

Let's look at the `ProtectedServlet` first. This is the class that defines the OIDC annotation and will redirect to Auth0 to handle OIDC authentication.

`src/main/java/com/demo/ProtectedServlet.java`

```java
package com.demo;

...

@OpenIdAuthenticationMechanismDefinition(
        providerURI = "${openIdConfig.issuerUri}",
        clientId = "${openIdConfig.clientId}",
        clientSecret = "${openIdConfig.clientSecret}",
        redirectURI = "${baseURL}/callback",
        // default 500ms caused timeouts for me
        jwksConnectTimeout = 5000,
        jwksReadTimeout = 5000,
        // Auth0 requires the audience to be set to the custom API

>         extraParameters = {"audience=http://my-api"},

Reviewer: Why are all the rest able to be properties and this value is hard-coded?

                Jwk jwk = provider.get(jwt.getKeyId());

                Algorithm algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);

                JWTVerifier verifier = JWT.require(algorithm)
                        .withIssuer(openIdConfig.getIssuerUri())
                        .build();

                jwt = verifier.verify(accessToken);
                LOGGER.info("JWT decoded. sub=" + jwt.getClaims().get("sub"));
                request.setAttribute("jwt", jwt);

            } 
            
            ...

        }

        chain.doFilter(request, response);
    }

    @Override
    public void destroy() {
    }

}
```

This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT verification](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute and the request is allowed to continue.

Give it a try. Start the project.

```bash
./mvnw wildfly:run
```

Make a request to the protected API endpoint (not the OIDC endpoint).

```bash
http :8080/api/protected
```

You'll get:

```Bash
HTTP/1.1 401 Unauthorized
```

> Now, use your OIDC endpoint to retrieve a token. Using a browser, open [http://localhost:8080/protected](http://localhost:8080/protected)

Reviewer: ```suggestion
Now, use your OIDC endpoint to retrieve a token. Using a browser, open `http://localhost:8080/protected`.
```

> Authenticate with Auth0. When you are redirected back to the protected servlet page, copy the token value and save it in a Bash shell variable in a new Bash shell.

Reviewer: ```suggestion
Authenticate with Auth0. When you are redirected back to the protected servlet page, copy the token value and save it as a variable in a new shell.
```

```bash
TOKEN=eyJraWQiOiJqY3dpbGpUcGVZSG1Jajl6ODR3LV...
```

In that same shell, make a request to the protected API endpoint using the token.

```bash
http :8080/api/protected "Authorization: Bearer $TOKEN"
```

It should return something like the following.

```bash
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 401
Content-Type: text;charset=ISO-8859-1
Date: Tue, 27 Sep 2022 15:02:59 GMT

Welcome, andrew.hughes@mail.com
...

```

## Keep Learning with Jakarta EE and Auth0

> You just built a Jakarta Enterprise Edition application that used the new Open ID connect annotation and implementation built into Jakarta EE 10. You used Auth0 as the OIDC and OAuth 2 provider, and you saw how to implement both SSO and JWT authentication.

Reviewer: ```suggestion
You just built a Jakarta Enterprise Edition application that used the new Open ID connect annotation and implementation built into Jakarta EE 10. You used Auth0 as the OIDC and OAuth 2.0 provider, and you saw how to implement both SSO and JWT authentication.
```

> ou can find the source code for this example on GitHub in the [@oktadev/okta-spring-boot-vue-crud-example](https://github.com/oktadev/okta-spring-boot-vue-crud-example) repository.

Reviewer: ```suggestion
You can find the source code for this example on GitHub in the [@oktadev/okta-jakartaee-oidc-example](https://github.com/oktadev/okta-jakartaee-oidc-example) repository.
```
